caddyhttp/httpserver/replacer.go

```diff
@@ -132,6 +147,36 @@ func NewReplacer(r *http.Request, rr *ResponseRecorder, emptyValue string) Repla
 	return rep
 }
 
+func canLogRequest(r *http.Request) (canLog bool) {
+	if r.Method == "POST" || r.Method == "PUT" {
+		for _, cType := range r.Header[headerContentType] {
+			// the cType could have charset and other info
+			if strings.Index(cType, contentTypeJSON) > -1 || strings.Index(cType, contentTypeXML) > -1 {
+				canLog = true
+				break
+			}
+		}
+	}
+	return
+}
+
+// readRequestBody reads the request body and sets a
+// new io.ReadCloser that has not yet been read.
+func readRequestBody(r *http.Request, n int64) ([]byte, error) {
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, n))
```

Comment:
Oh, good point. We don't need to penalize the client in this case, just read the first 100 KB at most with no other side-effects. This works as-is.